client/client.go

```diff
@@ -99,56 +123,513 @@ func UploadAdd(issuer principal.Signer, space did.DID, params uploadadd.Caveat,
+				concludeFxs = append(concludeFxs, inv)
+			case httpcap.PutAbility:
+				putTask = inv
+			case blob.AcceptAbility:
+				acceptTask = inv
+			case w3sBlob.AcceptAbility:
+				if acceptTask == nil {
+					acceptTask = inv
+					legacyAccept = true
+				}
+			}
+		}
+	}
+
+	if allocateTask == nil || len(concludeFxs) == 0 || putTask == nil || acceptTask == nil {
+		return nil, nil, fmt.Errorf("mandatory tasks not received in space/blob/add receipt")
+	}
+
+	var allocateRcpt receipt.Receipt[blob.AllocateOk, fdm.FailureModel]
+	var legacyAllocateRcpt receipt.Receipt[w3sBlob.AllocateOk, fdm.FailureModel]
+	var putRcpt receipt.AnyReceipt
+	var acceptRcpt receipt.Receipt[blob.AcceptOk, fdm.FailureModel]
+	var legacyAcceptRcpt receipt.Receipt[w3sBlob.AcceptOk, fdm.FailureModel]
+	for _, concludeFx := range concludeFxs {
+		concludeRcpt, err := getConcludeReceipt(concludeFx)
+		if err != nil {
+			return nil, nil, fmt.Errorf("reading ucan/conclude receipt: %w", err)
+		}
+
+		switch concludeRcpt.Ran().Link() {
+		case allocateTask.Link():
+			switch allocateTask.Capabilities()[0].Can() {
+			case blob.AllocateAbility:
+				allocateRcpt, err = receipt.Rebind[blob.AllocateOk, fdm.FailureModel](concludeRcpt, blob.AllocateOkType(), fdm.FailureType(), captypes.Converters...)
+				if err != nil {
+					return nil, nil, fmt.Errorf("bad allocate receipt in conclude fx: %w", err)
+				}
+			case w3sBlob.AllocateAbility:
+				legacyAllocateRcpt, err = receipt.Rebind[w3sBlob.AllocateOk, fdm.FailureModel](concludeRcpt, w3sBlob.AllocateOkType(), fdm.FailureType(), captypes.Converters...)
+				if err != nil {
+					return nil, nil, fmt.Errorf("bad allocate receipt in conclude fx: %w", err)
+				}
+			}
+		case putTask.Link():
+			putRcpt = concludeRcpt
+		case acceptTask.Link():
+			switch allocateTask.Capabilities()[0].Can() {
+			case blob.AcceptAbility:
+				acceptRcpt, err = receipt.Rebind[blob.AcceptOk, fdm.FailureModel](concludeRcpt, blob.AcceptOkType(), fdm.FailureType(), captypes.Converters...)
+				if err != nil {
+					return nil, nil, fmt.Errorf("bad accept receipt in conclude fx: %w", err)
+				}
+			case w3sBlob.AcceptAbility:
+				legacyAcceptRcpt, err = receipt.Rebind[w3sBlob.AcceptOk, fdm.FailureModel](concludeRcpt, w3sBlob.AcceptOkType(), fdm.FailureType(), captypes.Converters...)
+				if err != nil {
+					return nil, nil, fmt.Errorf("bad accept receipt in conclude fx: %w", err)
+				}
+			}
+		}
+	}
+
+	if allocateRcpt == nil && legacyAllocateRcpt == nil {
+		return nil, nil, fmt.Errorf("mandatory receipts not received in space/blob/add receipt")
+	}
+
+	var url url.URL
+	var headers http.Header
+	if allocateRcpt != nil {
+		allocateOk, err := result.Unwrap(result.MapError(allocateRcpt.Out(), failure.FromFailureModel))
+		if err != nil {
+			return nil, nil, fmt.Errorf("blob allocation failed: %w", err)
+		}
+
+		address := allocateOk.Address
+		if address == nil {
+			return nil, nil, fmt.Errorf("blob allocation failed: no address")
```

Comment:
I could be wrong, but I believe this is not correct

There is a legitimate reason address could be empty -- for a blob that is already uploaded. When this is the case, we should skip the upload entirely and right to accept.

@alanshaw can you confirm if I'm reading this right?

see: https://github.com/storacha/upload-service/blob/main/packages/upload-client/src/blob/add.js#L230

Reply:
changed to follow the implementation in `upload-client`. If this is not correct we can change it back afterwards.

Reply:
Yes, I can confirm, no address is the indication that the service already has the blob and no upload is necessary. It is not an error.

https://github.com/storacha/specs/blob/main/w3-blob.md#allocation-address